cpp/arcticdb/version/local_versioned_engine.cpp

```diff
@@ -1768,20 +1768,61 @@ timestamp LocalVersionedEngine::latest_timestamp(const std::string& symbol) {
 }
 
 std::unordered_map<KeyType, std::pair<size_t, size_t>> LocalVersionedEngine::scan_object_sizes() {
-    std::unordered_map<KeyType, std::pair<size_t, size_t>> sizes;
+    std::unordered_map<KeyType, std::pair<size_t, std::atomic<size_t>>> sizes;
     foreach_key_type([&store=store(), &sizes=sizes](KeyType key_type) {
-        std::vector<std::pair<VariantKey, stream::StreamSource::ReadContinuation>> keys_and_continuations;
-        auto& pair = sizes[key_type];
-        store->iterate_type(key_type, [&keys_and_continuations, &pair](const VariantKey &&k) {
-            keys_and_continuations.emplace_back(
-                std::forward<const VariantKey>(k),
-                [&pair](auto &&ks) {
-                    auto key_seg = std::move(ks);
-                    ++pair.first;
-                    pair.second += key_seg.segment().total_segment_size();
-                    return key_seg.variant_key();
-                });
+        std::vector<std::pair<VariantKey, stream::StreamSource::ReadContinuation>> keys;
+        store->iterate_type(key_type, [&keys, &sizes, &key_type](const VariantKey&& k) {
+            auto& pair = sizes[key_type];
+            ++pair.first;
+            auto& size_counter = pair.second;
+            keys.emplace_back(std::forward<const VariantKey>(k), [&size_counter] (auto&& ks) {
+                auto key_seg = std::move(ks);
+                size_counter += key_seg.segment().total_segment_size();
```

Comment:
There's an element of 'no good deed goes unpunished', but since we're changing this (and adding a test, which should have been there from the outset - my bad), I believe we're scanning the encoded fields and working out both the uncompressed and the compressed sizes. Do you think it would be possible to record both? I'm very interested in the compression ratios we're getting as I think we may want to move away from block encoders like LZ4

Reply:
This is all recording compressed sizes. Is there a way to get the uncompressed size without actually doing the decompression of the segments in to `SegmentInMemory`?